### PR TITLE
[flang] Lower REDUCE intrinsic with DIM argument

### PR DESCRIPTION
**Pull request by clementval**
This is a follow up patch to #94652 and handles the lowering of the reduce intrinsic with DIM argument and non scalar result.

### CODE REVIEW

File: flang/lib/Optimizer/Builder/Runtime/Reduction.cpp [hunk 1]
```diff
@@ -527,10 +571,32 @@ struct ForcedReduceComplex10 {
...
+          {});
+    };
+  }
+};
+
 /// Placeholder for complex(16) version of Reduce Intrinsic
 struct ForcedReduceComplex16 {
   static constexpr const char *name =
-      ExpandAndQuoteKey(RTNAME(CppReduceComplex16));
+      ExpandAndQuoteKey(RTNAME(CppReduceComplex16Dim));
```
Comment (vzakhari) [inline flang/lib/Optimizer/Builder/Runtime/Reduction.cpp:599]: Is this change intended?

Comment (clementval) [reply flang/lib/Optimizer/Builder/Runtime/Reduction.cpp:599]: No bad change! Thanks for catching it!

Comment (clementval) [inline flang/lib/Optimizer/Builder/Runtime/Reduction.cpp:599]:
```suggestion
      ExpandAndQuoteKey(RTNAME(CppReduceComplex16));
```

File: flang/lib/Optimizer/Builder/Runtime/Reduction.cpp [hunk 2]
```diff
@@ -540,9 +606,31 @@ struct ForcedReduceComplex16 {
       auto strTy = fir::ReferenceType::get(mlir::IntegerType::get(ctx, 8));
       auto intTy = mlir::IntegerType::get(ctx, 8 * sizeof(int));
       auto refTy = fir::ReferenceType::get(ty);
+      auto refBoxTy = fir::ReferenceType::get(boxTy);
```
Comment (vzakhari) [inline flang/lib/Optimizer/Builder/Runtime/Reduction.cpp:609]: Then, I guess, this should not be needed as well :)

Comment (clementval) [reply flang/lib/Optimizer/Builder/Runtime/Reduction.cpp:609]: This is needed because it was actually not tested in the previous patch. I can move this change to the other patch since it's not merged yet.

Comment (vzakhari) [reply flang/lib/Optimizer/Builder/Runtime/Reduction.cpp:609]: Hmm, I must be missing something.

The runtime API is defined as:
```
void RTDEF(CppReduceComplex16)(std::complex<CppFloat128Type> &result,
    const Descriptor &array,
    ReductionOperation<std::complex<CppFloat128Type>> operation,
    const char *source, int line, int dim, const Descriptor *mask,
    const std::complex<CppFloat128Type> *identity, bool ordered) {
```
So the first argument should be a reference to a complex value, isn't it right?

File: flang/lib/Optimizer/Builder/Runtime/Reduction.cpp [hunk 3]
```diff
@@ -540,9 +606,31 @@ struct ForcedReduceComplex16 {
...
+      auto ty = mlir::ComplexType::get(mlir::FloatType::getF128(ctx));
+      auto boxTy =
+          fir::runtime::getModel<const Fortran::runtime::Descriptor &>()(ctx);
+      auto opTy = mlir::FunctionType::get(ctx, {ty, ty}, ty);
+      auto strTy = fir::ReferenceType::get(mlir::IntegerType::get(ctx, 8));
+      auto intTy = mlir::IntegerType::get(ctx, 8 * sizeof(int));
+      auto refTy = fir::ReferenceType::get(ty);
+      auto i1Ty = mlir::IntegerType::get(ctx, 1);
+      return mlir::FunctionType::get(
+          ctx, {boxTy, boxTy, opTy, strTy, intTy, intTy, boxTy, refTy, i1Ty},
```
Comment (vzakhari) [inline flang/lib/Optimizer/Builder/Runtime/Reduction.cpp:633]: The first argument is a box reference here but not in the non-dim case.

Comment (clementval) [reply flang/lib/Optimizer/Builder/Runtime/Reduction.cpp:633]: Let me double check that. I have probably messade up smth here

File: flang/lib/Optimizer/Builder/Runtime/Reduction.cpp (hunk 2 above)
Comment (clementval) [reply flang/lib/Optimizer/Builder/Runtime/Reduction.cpp:609]: Yes you are right. I messed up the this one with the DIM version. I'll revert it back.

**Comment by clementval** [comment]
Thanks Slava and Jean for the review! I fixed the mismatch in `ForcedReduceComplex16` and `ForcedReduceComplex16Dim`. Thanks Slava for catching that.

**Review by vzakhari — APPROVED** [review]
Thank you, Valentin!

**Review by jeanPerier — APPROVED** [review]
LGTM

**Comment by jeanPerier** [comment]
Beware that the windows failure seems related to this change or the previous one: 
```
C:\ws\src\flang\lib\Optimizer\Builder\Runtime\Reduction.cpp(1499): error C2065: '_FortranAReduceReal10': undeclared identifier
```

I think you need "Forced" definition for the REAL*10 too.

**Comment by clementval** [comment]
> Beware that the windows failure seems related to this change or the previous one: 
> 
> ```
> 
> C:\ws\src\flang\lib\Optimizer\Builder\Runtime\Reduction.cpp(1499): error C2065: '_FortranAReduceReal10': undeclared identifier
> 
> ```
> 
> 
> 
> I think you need "Forced" definition for the REAL*10 too.

I'll look at that! Thanks

**Review by psteinfeld — APPROVED** [review]
All builds and tests correctly and looks good.
